pkg/virt-controller/watch/vm.go

```diff
@@ -428,41 +429,150 @@ func (c *VMController) handleDataVolumes(vm *virtv1.VirtualMachine, dataVolumes
-			if vmi.IsFinal() {
-				// The VirtualMachineInstance can fail od be finished. The job of this controller
+			forceRestart := false
+			if vm.Status.StateChangeRequest != nil &&
+				vm.Status.StateChangeRequest.StopVMI != nil &&
+				*vm.Status.StateChangeRequest.StopVMI == vmi.UID {
+				forceRestart = true
+			}
+			if forceRestart || vmi.IsFinal() {
+				// The VirtualMachineInstance can fail or be finished. The job of this controller
 				// is keep the VirtualMachineInstance running, therefore it restarts it.
 				// restarting VirtualMachineInstance by stopping it and letting it start in next step
 				err := c.stopVMI(vm, vmi)
 				if err != nil {
 					log.Log.Object(vm).Error("Cannot restart VirtualMachineInstance, the VirtualMachineInstance cannot be deleted.")
-					return err
+					return false, err
 				}
 				// return to let the controller pick up the expected deletion
 			}
 			// VirtualMachineInstance is OK no need to do anything
-			return nil
+			return false, nil
 		}
 
 		err := c.startVMI(vm)
-		return err
-	}
+		if err != nil {
+			return false, err
+		}
+		return true, nil
+
+	case virtv1.RunStrategyRestartOnError:
+		// For this RunStrategy, a VMI should always be restarted if it failed.
+		// If a VMI enters the Succeeded phase, it should not be restarted.
+		if vmi != nil {
+			forceStop := false
+			// If there's a stop request that matches the existing VMI's UUID
+			if vm.Status.StateChangeRequest != nil &&
+				vm.Status.StateChangeRequest.StopVMI != nil &&
+				*vm.Status.StateChangeRequest.StopVMI == vmi.UID {
+				forceStop = true
+			}
+
+			// If the existing VMI failed
+			if vmi.Status.Phase == virtv1.Failed {
+				// Patch (etcd's view of) the VM on the spot. This controller can't rely on deleting the VMI to
+				// restart it (because a failed-then-deleted and a succeeded-then-deleted VMI are both deleted)
+				bodyString := fmt.Sprintf("{\"status\": {\"stateChangeRequest\": {\"startVMI\": true}}}")
+				_, err := c.clientset.VirtualMachine(vm.ObjectMeta.Namespace).Patch(vm.GetName(), types.MergePatchType, []byte(bodyString))
+				if err != nil {
+					log.Log.Object(vm).Error("Cannot restart VirtualMachineInstance, the VirtualMachine cannot be patched.")
+					return false, err
+				}
+			}
+
+			if forceStop || vmi.IsFinal() {
+				// For RestartOnError, this controller should only restart the VirtualMachineInstance
+				// if it failed.
+				err := c.stopVMI(vm, vmi)
+				if err != nil {
+					log.Log.Object(vm).Error("Cannot restart VirtualMachineInstance, the VirtualMachineInstance cannot be deleted.")
+					return false, err
+				}
+				// return to let the controller pick up the expected deletion
+			}
+			// VirtualMachineInstance is OK no need to do anything
+			return false, nil
+		}
+
+		startVMI := false
+		if vm.Status.StateChangeRequest != nil {
+			startVMI = vm.Status.StateChangeRequest.StartVMI
+		}
+
+		// FIXME: double check if status.created can be used to start VMI initially
```

Comment:
Ah, now I understand the problem: how to determine if there was a failed VMI before.
Well, I would say we need a solution for this, because I would really expect a initial start of the VMI with this run strategy, I noticed this when reading the tests, and I was wondering why you start VMs there.
I think I would keep track of what happened (was there a failed VMI?) in the status of the VM.